monkey/infection_monkey/master/ip_scanner.py

```diff
@@ -0,0 +1,69 @@
+import logging
+import queue
+import threading
+from queue import Queue
+from threading import Event
+from typing import Callable, Dict, List
+
+from infection_monkey.i_puppet import IPuppet, PingScanData, PortScanData
+
+from .threading_utils import create_daemon_thread
+
+logger = logging.getLogger()
+
+Callback = Callable[[str, PingScanData, Dict[int, PortScanData]], None]
+
+
+class IPScanner:
+    def __init__(self, puppet: IPuppet, num_workers: int):
+        self._puppet = puppet
+        self._num_workers = num_workers
+
+    def scan(self, ips_to_scan: List[str], options: Dict, results_callback: Callback, stop: Event):
+        # Pre-fill a Queue with all IPs to scan so that threads know they can safely exit when the
+        # queue is empty.
+        ips = Queue()
+        for ip in ips_to_scan:
+            ips.put(ip)
+
+        scan_ips_args = (ips, options, results_callback, stop)
+        scan_threads = []
+        for i in range(0, self._num_workers):
+            t = create_daemon_thread(target=self._scan_ips, args=scan_ips_args)
+            t.start()
+            scan_threads.append(t)
+
+        for t in scan_threads:
+            t.join()
+
+    def _scan_ips(self, ips: Queue, options: Dict, results_callback: Callback, stop: Event):
+        logger.debug(f"Starting scan thread -- Thread ID: {threading.get_ident()}")
+
+        try:
+            while not stop.is_set():
+                ip = ips.get_nowait()
+                logger.info(f"Scanning {ip}")
+
+                ping_scan_data = self._puppet.ping(ip, options["icmp"])
+                port_scan_data = self._scan_tcp_ports(ip, options["tcp"], stop)
+
+                results_callback(ip, ping_scan_data, port_scan_data)
```

Comment:
Maybe it would be more readable if this class wouldn't use a callback. Instead, we would put all of the results into a data structure and the user (who actually knows what the callback is) would pull the data from the data structure and do whatever. In this case the `Propagator._scan_network` would do something like:
```
while not stop:
    if len(self._ip_scanner.data) > 0:
        for ip_scan_result in self._ip_scanner.data:
            self._process_scan_results(ip_scan_results)
```

Reply:
If you do that, then the `IPScanner` becomes responsible for controlling the flow of the `Propagator`. There needs to be a separate signaling mechanism for the scanner to signal that it has completed its work. Any bug in how that mechanism is managed results in flow control issues in the `Propagator`. For example, an uncaught exception could result in the propagator looping indefinitely while the scanner is no longer running.